Add rgb

cwpitts opened the pull request:
### Brief summary of what the package does

`rgb.el` uses [OpenRGB](https://gitlab.com/CalcProgrammer1/OpenRGB) to provide a Lisp interface for RGB device control.

### Direct link to the package repository

https://gitlab.com/cwpitts/rgb.el

### Your association with the package

I am the sole author and maintainer of `rgb.el`.

### Relevant communications with the upstream package maintainer

None needed.

### Checklist

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

riscy commented:
Hello - looks like the repo is not public.

cwpitts commented:
@riscy Fixed! You should be good to go.

cwpitts commented:
@riscy Just following up, is it possible to get another review of this PR?

riscy commented:
Thanks for your patience!  When you built and installed the package you may have noticed the following files are included in your recipe:

- el-mock.el (GPL*) -- Tiny Mock and Stub framework in Emacs Lisp
- ert-expectations.el (GPL*) -- The simplest unit test framework in the world
- rgb.el (MIT License) -- RGB control from Emacs.
- test-rgb.el (unknown license) -- (no header)

The recipe should be adjusted accordingly, but I assume you only want to be including rgb.el, so I will focus on just that:

`rgb.el` with package-lint 20220607.738:
```
6 issues found:
1:0: warning: "Version:" or "Package-Version:" header is missing. MELPA will handle this, but other archives will not.
1:0: warning: Including "Emacs" in the package summary is usually redundant.
1:0: error: Package should have a Homepage or URL header.
1:0: warning: The package summary should not end with a period.
30:10: error: You should depend on (emacs "24.3") or the cl-lib package if you need `cl-lib'.
34:1: error: You should depend on (emacs "24.3") or the cl-lib package if you need `cl-defun'.
```

However, on a (very) quick glance, this package seems to just be a single function that passes through arguments to a CLI program.  This may be useful but I think it's not quite innovative at this point in development.  (For example it would be questionable, in my opinion, for MELPA to have a package for every single CLI app in existence unless there were some drastic interface/workflow improvement -- which may be the case here, if you could demonstrate?)

cwpitts commented:
Thanks for getting back to me! I now get a clean report from `package-lint-current-buffer`, and I believe I've correctly adjusted the recipe.

This *is* a pretty simple package, mostly because it's intended to enable other things. I'm also working on [rgb-mode-lights.el](https://gitlab.com/cwpitts/rgb-mode-lights.el/-/blob/master/rgb-mode-lights.el), which enables a user to configure a hook that triggers on buffer changes. The hook detects the buffer's major mode and uses user-supplied configuration to set RGB device settings (color, mode, etc). I also plan to write more code that can check the result of a compilation command and set a keyboard or other device to red/green depending on the result of the command.

I considered putting all of this in the same package, but thought it would be better to separate the pure OpenRGB shim from the more functional code (for example, you might want to do something else with RGB devices, but not really want to configure the mode-change hook). If you think it would help to cross the threshold for a sufficiently innovative package, I'd be happy to combine all of them and set up flags for the user to turn different parts on or off. Which, now that I've written that out, I might do anyway, since that might be a nicer way to package things.

riscy commented:
> Which, now that I've written that out, I might do anyway, since that might be a nicer way to package things.

I think this would be the way to go for this package.  I can definitely see the utility here as a starting point though.

cwpitts commented:
Would the proper procedure from this point be to close this PR and open a new one once the updated package is ready? Or should I leave this open and tag you when I think it's ready to go?

riscy commented:
For sure, you're welcome to keep this one open and just tag me when it's ready to go and I'll be happy to follow up.  I'll mark it awaiting-upstream in the meantime.

cwpitts commented:
@riscy Ready for another look! Major design changes:

1. Moved all code into `rgb.el`, `rgb-mode-lights.el` will be deprecated and archived.
2. Added support for different backends to map config to calls to a RGB control program (default is OpenRGB). If someone has hardware that doesn't work with OpenRGB, they can now write a custom backend that handles argument marshaling and calling the correct program. All of this is documented in the `README.md` file.

`package-lint` and `checkdoc` are still happy after the refactor, and tests still pass.

riscy commented on recipes/rgb:
```diff
@@ -0,0 +1,4 @@
+(rgb
+ :fetcher gitlab
+ :repo "cwpitts/rgb.el"
+ :files ("rgb.el"))
```
I'm not saying you must, but having `:defaults` in the recipe gives you room to grow and is slightly preferable:
```suggestion
(rgb
 :fetcher gitlab
 :repo "cwpitts/rgb.el"
 :files (:defaults
         (:exclude "el-mock.el"
                   "ert-expectations.el"
                   "test-rgb.el")))
```
And if you renamed test-rgb.el to rgb-test.el, it will automatically be excluded by `:defaults`, and in that case you could write:
```suggestion
(rgb
 :fetcher gitlab
 :repo "cwpitts/rgb.el"
 :files (:defaults
         (:exclude "el-mock.el" "ert-expectations.el"))
```

cwpitts replied:
Suggestions applied!

riscy reviewed (commented):
Thanks, this looks good.  Just some quick comments on the new code, but nothing too tricky:

- See the code suggestion re: the recipe
- rgb.el#L50: It's safer to sharp-quote function names; use `#'`
- The autoload cookie `;;###autoload` should have three semicolons I believe `;;;###autoload`
- Rather than a function called `rgb-enable-mode-change-hook` have you considered implementing a global minor mode called `rgb-mode`?  That would allow users to toggle it off again with <kbd>M-x rgb-mode</kbd>

riscy commented:
I just noticed this too, but you're using triple semicolons in your front matter where for the most part you want to be using double semicolons:

```
;;; rgb.el --- RGB control via OpenRGB
;; Version: 2.0.0
;; URL: https://gitlab.com/cwpitts/rgb.el
;; Package-Requires: ((emacs "24.3"))

;; Copyright © 2022 Christopher Pitts <cwpitts@protonmail.com>
;; ...

;;; Commentary:
;; This package uses RGB CLI tools to control RGB devices
;; ...

;;; Code:
```
This is because the triple semicolons are for section headings.

cwpitts commented:
Applied your suggestion on the packaging function, and the adjusted comment prefixes are now in `master`! I successfully re-ran both `make recipes/rgb` and `STABLE=t make recipes/rgb`. I note that the CI pipeline failed, but I'm not sure how to interpret the "end of file during parsing" error that's displayed.

cwpitts commented:
RE: Global minor mode - I've thought about it, but I haven't implemented minor modes before. I"ll read up on it and decide if that's the right way to go for this package.

cwpitts commented:
After a perusal of the documentation, I decided to implement it that way (changes now on `master`). I also found and fixed the issue causing CI to fail (naturally, it was a missing close parenthesis in the `rgb` recipe).

riscy commented:
I think this is a positive direction.  One thing I think you should do is add and remove the hook as part of the minor mode, rather than having it always active (and always checking the value of a variable).

You could probably just name the mode `rgb-mode` -- since there's no local variant of the mode -- and specifying a group will alleviate the byte-compile warnings:
```
(defgroup rgb nil
  "Docstring..."
  :prefix "rgb-"
  :group 'games)
```
(You may want to peruse the manual or google for appropriate parent groups)

cwpitts commented:
I was wondering about my implementation; I didn't like the idea of having a minor mode "leaking", so to speak, and having a persistent hook function even when it's not active. I added the hook toggling and customization group, `checkdoc` and `package-lint` are still happy and a test MELPA build passes locally! For the group I decided on `External`, since RGB hardware is by definition external to Emacs itself.

riscy commented:
That looks good to me!  One more thing I notice is that when you use `define-minor-mode` it will automatically create a variable that you can use to check whether the mode is on or off, and so you can replace:

```elisp
(define-minor-mode rgb-mode
  "Global RGB mode."
  :global t
  :lighter " RGB"
  :group 'rgb
  (if (bound-and-true-p global-rgb-mode)
      (add-hook 'window-buffer-change-functions
                #'rgb-window-buffer-change-hook)
    (remove-hook 'window-buffer-change-functions
                 #'rgb-window-buffer-change-hook)))
```
with:
```elisp
(define-minor-mode rgb-mode
  "Global RGB mode."
  :global t
  :lighter " RGB"
  :group 'rgb
  (if rgb-mode
      (add-hook 'window-buffer-change-functions
                #'rgb-window-buffer-change-hook)
    (remove-hook 'window-buffer-change-functions
                 #'rgb-window-buffer-change-hook)))
```

cwpitts commented:
Thanks for catching that! I missed renaming that variable when I renamed the mode to `rgb-mode`. Fix pushed to `master`!

riscy commented:
I apologize for the back and forth but I looked into the code a little more deeply (which is a good thing, it means I'm interested :) and I'm confused by:

`(if mode (set 'args (append args (list "-m" (format "%s" mode)))))`

where I *think* that `mode` is meant to specify an [OpenRGB "device mode"](https://github.com/CalcProgrammer1/OpenRGB/blob/master/cli.cpp#L392=), but if I trace backward through the code we seem to be passing in an Emacs lisp major mode (e.g. "emacs-lisp-mode" or "org-mode" etc.)  I may be completely out to lunch and misreading that `-m` flag, but I wanted to bring it up before merging.

Some other quick stuff --

- Rather than documenting an unused variable ("WINDOW parameter not used.") you can just call that variable `_window` and it will be ignored
- This is of course up to you but I wonder if `window-state-change-functions` would be preferable to `window-buffer-change-functions` as the landing place for your hook, since moving around between my frames wouldn't trigger openrgb updates

cwpitts commented:
No need to apologize! You're correct that for the OpenRGB device backend that parameter is the device mode. It was a poor choice of variable name on my part (the fact that my tests passed and _I_ understood what was happening blinded me to the fact that someone else might not realize I was using two different `mode` variables). I changed that parameter to `light-mode` in the function to remove any ambiguity. For good measure I also renamed the `mode` variable used elsewhere that holds the current major mode to be `current-major-mode`.

Looking at the window hooks documentation again, I noticed that `window-state-change-functions` is specifically mentioned as being preferred for detecting state changes (from the docs on `window-state-change-hook`:

> Applications should put a function on this hook only if they want to react to changes that happened on (or have been signaled for) two or more frames since last redisplay. In every other case, putting the function on window-state-change-functions should be preferred.

Both seem to work, and I'm pretty agnostic on which one, so I changed it to be a state change hook!

riscy commented:
Thanks, that was my mistake re: the mode.  Merging. :)
